src/data_source/sql.rs

```diff
@@ -0,0 +1,987 @@
+            tx_in_progress: false,
+            kill: Some(kill),
+        })
+    }
+
+    /// Query the underlying SQL database.
+    pub async fn query<T, P>(&self, query: &T, params: P) -> Result<RowStream, Error>
+    where
+        T: ?Sized + ToStatement,
+        P: IntoIterator,
+        P::IntoIter: ExactSizeIterator,
+        P::Item: BorrowToSql,
+    {
+        Ok(self.client.query_raw(query, params).await?)
+    }
+
+    /// Access the transaction which is accumulating all uncommitted changes to the data source.
+    ///
+    /// This can be used to manually group database modifications to custom state atomically with
+    /// modifications made through the [`SqlDataSource`].
+    ///
+    /// If there is no currently open transaction, a new transaction will be opened. No changes
+    /// made through the transaction objeect returned by this method will be persisted until
+    /// [`commit_version`](Self::commit_version) is called.
+    pub async fn transaction(&mut self) -> Result<Transaction<'_>, Error> {
+        if !self.tx_in_progress {
+            // If there is no transaction in progress, open one.
+            self.client.batch_execute("BEGIN").await?;
+            self.tx_in_progress = true;
+        }
+        Ok(Transaction {
+            client: &mut self.client,
+        })
+    }
+}
+
+impl Drop for SqlDataSource {
+    fn drop(&mut self) {
+        if let Some(kill) = self.kill.take() {
+            // Ignore errors, they just mean the task has already exited.
+            kill.send(()).ok();
+        }
+    }
+}
+
+#[async_trait]
+impl VersionedDataSource for SqlDataSource {
+    type Error = postgres::error::Error;
+
+    /// Atomically commit to all outstanding modifications to the data.
+    ///
+    /// If this method fails, outstanding changes are left unmodified. The caller may opt to retry
+    /// or to erase outstanding changes with [`revert_version`](Self::revert_version).
+    async fn commit_version(&mut self) -> Result<(), Self::Error> {
+        if !self.tx_in_progress {
+            // Nothing to do if there is no outstanding transaction.
+            return Ok(());
+        }
+        self.client.batch_execute("COMMIT").await?;
+        self.tx_in_progress = false;
+        Ok(())
+    }
+
+    /// Erase all oustanding modifications to the data.
+    ///
+    /// This function must not return if it has failed to revert changes. Inability to revert
+    /// changes to the database is considered a fatal error, and this function may panic.
+    async fn revert_version(&mut self) {
+        if !self.tx_in_progress {
+            // Nothing to do if there is no outstanding transaction.
+            return;
+        }
+        // If we're trying to roll back a transaction, something has already gone wrong and we're
+        // trying to recover. If we're unable to revert the changes and recover, all we can do is
+        // panic.
+        self.client.batch_execute("ROLLBACK").await.unwrap();
```

Comment:
Something like `expect("DB rollback succeeds.")` might help the person looking at the panic.